src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java

```diff
@@ -0,0 +1,108 @@
+ * relevant to the code. This includes for example the `else` expression and curly braces.
+ * For each {@link ChangeBlock}:
+ * Check if there are such lines before or after the ChangeBlock. If there are add the entire prefix/postfix as a
+ * potential ChangeBlock to the GroupedFile.
+ */
+public class AddUncoveredLinesAsPotentialCodeBlocks extends BehavioralKnowledgeSource {
+
+    private static final Pattern CURLY_BRACES_PATTERN = Pattern.compile("\\s*[}{]\\s*");
+
+    private static final Pattern ELSE_PATTERN = Pattern.compile("\\s*}?\\s*else\\s*\\{?\\s*");
+
+    private static final Pattern EMPTY_LINE_PATTERN = Pattern.compile("\\s*");
+
+    public AddUncoveredLinesAsPotentialCodeBlocks(BehavioralBlackboard blackboard) {
+        super(blackboard);
+    }
+
+    @Override
+    public boolean executeCondition() {
+        return blackboard.getGroupedFiles() != null && blackboard.getGroupedFiles().stream().noneMatch(groupedFile -> groupedFile.getFileContent() == null)
+                && blackboard.getGroupedFiles().stream().noneMatch(groupedFile -> groupedFile.getCommonChanges() == null);
+    }
+
+    @Override
+    public boolean executeAction() {
+        boolean didChanges = false;
+
+        for (GroupedFile groupedFile : blackboard.getGroupedFiles()) {
+            var newChangeBlocks = new TreeSet<ChangeBlock>();
+
+            for (ChangeBlock commonChange : groupedFile.getCommonChanges()) {
+                var firstLine = commonChange.getLines().first();
+                var potentialPrefix = getPotentialPrefix(firstLine, groupedFile.getFileContent());
+                if (potentialPrefix != null) {
+                    newChangeBlocks.add(potentialPrefix);
+                }
+                var lastLine = commonChange.getLines().last();
+                var potentialPostfix = getPotentialPostfix(lastLine, groupedFile.getFileContent());
+                if (potentialPostfix != null) {
+                    newChangeBlocks.add(potentialPostfix);
+                }
+            }
+
+            if (!newChangeBlocks.isEmpty()) {
+                groupedFile.getCommonChanges().addAll(newChangeBlocks);
+                didChanges = true;
+            }
+        }
+
+        return didChanges;
+    }
+
+    private ChangeBlock getPotentialPrefix(int firstLine, String fileContent) {
+        var potentialLines = new TreeSet<Integer>();
+        var lineContents = fileContent.split("\n");
+        for (int i = firstLine - 2; i >= 0; i--) {
+            var lineContent = lineContents[i];
+            if (doesLineMatch(lineContent)) {
+                potentialLines.add(i + 1);
+            }
+            else {
+                break;
+            }
+        }
+        if (potentialLines.isEmpty()) {
+            return null;
+        }
+        else {
+            return new ChangeBlock(potentialLines, true);
+        }
+    }
+
+    private ChangeBlock getPotentialPostfix(int lastLine, String fileContent) {
```

Comment:
Aside from the for-loop condition, all lines are equal for `getPotentialPrefix` and `getPotentialPostfix`.

A possibility to resolve this could be to define a private interface with a method that returns the line number (index) if the condition is still valid and -1 if the condition is not fulfilled anymore? -1 can never be a valid value for line numbers and therefore could be used for this.
The functionality could be extracted into a new method that additionally has this interface as a parameter. In `getPotentialPrefix` and `getPotentialPostfix` you could then hand over a lamba function describing this modified condition behavior.
What do you think about that idea?

Reply:
This would make the code quite a lot more complicated IMO. So I would leave it as it is instead of introducing a new interface just for this.

Reply:
Fine by me 👍